Orange/classification/rules.py

```diff
@@ -0,0 +1,947 @@
+            Validation algorithm.
+        general_validator : Validator
+            Validation algorithm.
+
+        Returns
+        -------
+        rules : Rule list
+            First rules developed in the process of learning a single
+            rule.
+        """
+        raise NotImplementedError("descendants of SearchStrategy must "
+                                  "overload method 'initialise_rule'")
+
+    def refine_rule(self, X, Y, candidate_rule):
+        """
+        Refine rule.
+
+        Parameters
+        ----------
+        X, Y : array_like
+            Learning data.
+        candidate_rule : Rule
+            Refine this rule.
+
+        Returns
+        -------
+        rules : Rule list
+            Descendant rules of 'candidate_rule'.
+        """
+        raise NotImplementedError("descendants of SearchStrategy must "
+                                  "overload method 'refine_rule'")
+
+
+class TopDownSearchStrategy(SearchStrategy):
+    """
+    An empty starting rule that covers all instances is developed and
+    added to the list of candidate rules. The hypothesis space of
+    possible rules is searched by repeatedly specialising candidate
+    rules.
+    """
+    def initialise_rule(self, X, Y, target_class, base_rules, domain,
+                        prior_class_dist, quality_evaluator,
+                        complexity_evaluator, significance_validator,
+                        general_validator):
+        rules = []
+        default_rule = Rule(domain=domain, prior_class_dist=prior_class_dist,
+                            quality_evaluator=quality_evaluator,
+                            complexity_evaluator=complexity_evaluator,
+                            significance_validator=significance_validator,
+                            general_validator=general_validator)
+
+        default_rule.filter_and_store(X, Y, target_class)
+        rules.append(default_rule)
+
+        for base_rule in base_rules:
+            temp_rule = Rule(selectors=copy(base_rule.selectors),
+                             parent_rule=default_rule, domain=domain,
+                             prior_class_dist=prior_class_dist,
+                             quality_evaluator=quality_evaluator,
+                             complexity_evaluator=complexity_evaluator,
+                             significance_validator=significance_validator,
+                             general_validator=general_validator)
+
+            temp_rule.filter_and_store(X, Y, target_class)
+            if temp_rule.validity:
+                rules.append(temp_rule)
+
+        return rules
+
+    def refine_rule(self, X, Y, candidate_rule):
+        covered_X = X[candidate_rule.covered_examples]
+        possible_selectors = []
+
+        for i, attribute in enumerate(candidate_rule.domain.attributes):
+            column = covered_X[:, i]
+            if attribute.is_discrete:
+                for val in [int(x) for x in set(column)]:
```

Comment:
`np.unique(column).astype(int)` will be much faster when `column` is already an array.